lib/WP_Auth0_LoginManager.php

```diff
@@ -163,136 +170,124 @@ public function init_auth0() {
-            $redirectURL = $state_decoded->redirect_to;
-          } else {
-            $redirectURL = $this->a0_options->get( 'default_login_redirection' );
-          }
+      // Look for clues as to what went wrong
+      throw new WP_Auth0_LoginFlowValidationException(
+        ! empty( $data->error_description ) ? $data->error_description : __( 'Unknown error', 'wp-auth0' ),
+        ( ! empty( $data->error ) ? $data->error : $exchange_resp_code )
+      );
+    }
 
-          wp_safe_redirect( $redirectURL );
-        }
-      }
-    } elseif ( is_array( $response['response'] ) &&  401 === (int) $response['response']['code'] ) {
+    $decoded_token = JWT::decode(
+      $data->id_token,
+      $this->a0_options->get_client_secret_as_key(),
+      array( $this->a0_options->get_client_signing_algorithm() )
+    );
 
-      $error = new WP_Error( '401', 'auth/token response code: 401 Unauthorized' );
+    $userinfo_resp = WP_Auth0_Api_Client::get_user(
+      $this->a0_options->get( 'domain' ),
+      WP_Auth0_Api_Client::get_client_token(),
+      $decoded_token->sub
+    );
 
-      WP_Auth0_ErrorManager::insert_auth0_error( __METHOD__ . ' => $this->login_user() = 401', $error );
+    $userinfo_resp_code = (int) wp_remote_retrieve_response_code( $userinfo_resp );
+    $userinfo_resp_body = wp_remote_retrieve_body( $userinfo_resp );
 
-      $msg = __( 'Error: the Client Secret configured on the Auth0 plugin is wrong. Make sure to copy the right one from the Auth0 dashboard.', 'wp-auth0' );
+    if ( $userinfo_resp instanceof WP_Error || 200 !== $userinfo_resp_code || empty( $userinfo_resp_body ) ) {
 
-      throw new WP_Auth0_LoginFlowValidationException( $msg );
-    } else {
-      $error = '';
-      $description = '';
+      // Management API call failed
+      // TODO: fallback to /userinfo
+      WP_Auth0_ErrorManager::insert_auth0_error( __METHOD__ . ' => WP_Auth0_Api_Client::get_user()', $userinfo_resp );
+      error_log( $userinfo_resp->get_error_message() );
+      throw new WP_Auth0_LoginFlowValidationException( __( 'Error getting user information', 'wp-auth0' ) );
+    }
 
-      if ( isset( $data->error ) ) {
-        $error = $data->error;
-      }
-      if ( isset( $data->error_description ) ) {
-        $description = $data->error_description;
-      }
+    $userinfo = json_decode( $userinfo_resp_body );
 
-      if ( ! empty( $error ) || ! empty( $description ) ) {
-        $error = new WP_Error( $error, $description );
-        WP_Auth0_ErrorManager::insert_auth0_error( __METHOD__ . ' => $this->login_user()', $error );
+    if ( $this->login_user( $userinfo, $data->id_token, $data->access_token ) ) {
+      $state_decoded = $this->get_state();
+      if ( ! empty( $state_decoded->interim ) ) {
+        include WPA0_PLUGIN_DIR . 'templates/login-interim.php';
+      } else {
+        if ( ! empty( $state_decoded->redirect_to ) && wp_login_url() !== $state_decoded->redirect_to ) {
```

Comment:
what about

```php
     $redirectURL= $state_decoded->redirect_to
      if ( empty( $redirectURL ) || wp_login_url() !==  $redirectURL ) {
         $redirectURL = $this->a0_options->get( 'default_login_redirection' );
       }
     wp_safe_redirect( $redirectURL );
```

Reply:
What if `$state_decoded->redirect_to` is not set?

Reply:
I guess that's not necessary to consider since I'm setting the state ...

Reply:
I don't know, I guess from your answer that it will throw an unhandled exception. If that's the case then fine 👍. I only suggested this as a way to avoid calling `$state_decoded->redirect_to` 3 times